### PR TITLE
Added Version

### PR DESCRIPTION
**Pull request by shanejonas**
Enables versioned deploy of GitHub Pages. See the updated README for details.

### CODE REVIEW

File: run.sh [hunk 1]
```diff
@@ -4,33 +4,76 @@ set -x
...
+  npx json -I -f package.json -e "this.homepage=this.homepage + \"${DESTINATION_DIRECTORY}/\""
+fi
+
 git config user.name github-actions
 git config user.email github-actions@github.com
 
-if git checkout --orphan gh-pages
+yarn "${PACKAGE_BUILD_COMMAND}"
+
+if git branch --list gh-pages
```
Comment (rekmarks) [inline run.sh]: Should this take a `--quiet` or `--exit-code` for the way we're using it?

File: run.sh [hunk 2]
```diff
@@ -4,33 +4,84 @@ set -x
...
-git worktree add "${DIRECTORY}" gh-pages
-yarn "${NPM_COMMAND}"
-cd "${DIRECTORY}"
-git add --all
-git commit -m "gh-pages deploy - ${NEW_VERSION}"
-git push -f origin gh-pages
\ No newline at end of file
+if [[ "$ADD" = "--add" ]]
+  then
+    npx json -I -f package.json -e "this.homepage=this.homepage + \"${DESTINATION_DIRECTORY}/\""
```
Comment (shanejonas) [inline run.sh]:
```suggestion
    # Edit package.json in place to reflect the new homepage url
    npx json -I -f package.json -e "this.homepage=this.homepage + \"${DESTINATION_DIRECTORY}/\""
```

File: run.sh [hunk 3]
```diff
@@ -4,33 +4,85 @@ set -x
...
+
+# If publishing directly to the root directory, erase all existing files on the
+# gh-pages branch. Otherwise, do not modify any files outside of the destination
+# directory.
+if [[ "$DESTINATION_DIRECTORY" = "." ]]; then
+  ADD=""
+else
+  # "--add" prevents existing files from being deleted.
+  ADD="--add"
+  # Edit package.json in place to reflect the new homepage url
```
Comment (shanejonas) [inline run.sh]:
```suggestion
```

File: package.json [hunk 4]
```diff
@@ -8,5 +8,6 @@
   "bugs": {
     "url": "https://github.com/MetaMask/action-publish-gh-pages/issues"
   },
-  "homepage": "https://github.com/MetaMask/action-publish-gh-pages#readme"
+  "homepage": "https://github.com/MetaMask/action-publish-gh-pages#readme",
+  "devDependencies": {}
```
Comment (rekmarks) [inline package.json]:
```suggestion
  "homepage": "https://github.com/MetaMask/action-publish-gh-pages#readme"
```

File: run.sh [hunk 5]
```diff
@@ -4,33 +4,84 @@ set -x
...
-cd "${DIRECTORY}"
-git add --all
-git commit -m "gh-pages deploy - ${NEW_VERSION}"
-git push -f origin gh-pages
\ No newline at end of file
+if [[ "$ADD" = "--add" ]]
+  then
+    # Edit package.json in place to reflect the new homepage url
+    # This is required by some build tools, e.g. Webpack
+    npx json -I -f package.json -e "this.homepage=this.homepage + \"${DESTINATION_DIRECTORY}/\""
```
Comment (shanejonas) [inline run.sh]:
```suggestion
    npx json -I -f package.json -e "this.homepage=(this.homepage.endsWith('/') ? this.homepage : this.homepage + "/") + \"${DESTINATION_DIRECTORY}/\""
```

Comment (rekmarks) [reply run.sh]:
```suggestion
    npx json -I -f package.json \
      -e "this.homepage=(this.homepage.replace(/\/$/u, '')) + \"/${DESTINATION_DIRECTORY}/\""
```

File: run.sh [hunk 6]
```diff
@@ -4,33 +4,85 @@ set -x
 set -e
 set -o pipefail
 
-DIRECTORY="${1}"
+SOURCE_DIRECTORY="${1}"
+RELEASE_BRANCH_PREFIX="${2}"
```
Comment (rekmarks) [inline run.sh]: Can we replace this input with the `NEW_VERSION` directly instead?

File: run.sh [hunk 7]
```diff
@@ -4,33 +4,74 @@ set -x
...
+fi
+
 git config user.name github-actions
 git config user.email github-actions@github.com
+git remote set-url origin "https://git:${GITHUB_TOKEN}@github.com/${GITHUB_REPOSITORY}.git"
+
+branch_exists=$(git ls-remote --quiet . origin/gh-pages)
 
-if git checkout --orphan gh-pages
+if [ -n "$branch_exists" ]
```
Comment (rekmarks) [inline run.sh]:
```suggestion
if [[ -n "$branch_exists" ]]
```

File: run.sh [hunk 8]
```diff
@@ -4,33 +4,74 @@ set -x
...
+else
+  # "--add" prevents existing files from being deleted.
+  ADD="--add"
+fi
+
 git config user.name github-actions
 git config user.email github-actions@github.com
+git remote set-url origin "https://git:${GITHUB_TOKEN}@github.com/${GITHUB_REPOSITORY}.git"
+
+branch_exists=$(git ls-remote --quiet . origin/gh-pages)
```
Comment (rekmarks) [inline run.sh]:
```suggestion

REPO="https://git:${GITHUB_TOKEN}@github.com/${GITHUB_REPOSITORY}.git"

git config user.name github-actions
git config user.email github-actions@github.com
git remote set-url origin "$REPO"

branch_exists=$(git ls-remote --quiet --heads "$REPO" gh-pages)
```

File: run.sh [hunk 9]
```diff
@@ -4,33 +4,76 @@ set -x
...
+REPO="https://git:${GITHUB_TOKEN}@github.com/${GITHUB_REPOSITORY}.git"
+
 git config user.name github-actions
 git config user.email github-actions@github.com
+git remote set-url origin "$REPO"
+
+branch_exists=$(git ls-remote --quiet --heads "$REPO" gh-pages)
 
-if git checkout --orphan gh-pages
+if [[ -n "$branch_exists" ]]
```
Comment (rekmarks) [inline run.sh:52]:
```suggestion
if [[ -n "$branch_exists" ]];
```

File: run.sh [hunk 10]
```diff
@@ -4,33 +4,76 @@ set -x
...
 fi
 
-git worktree add "${DIRECTORY}" gh-pages
-yarn "${NPM_COMMAND}"
-cd "${DIRECTORY}"
-git add --all
-git commit -m "gh-pages deploy - ${NEW_VERSION}"
-git push -f origin gh-pages
\ No newline at end of file
+if [[ "$ADD" = "--add" ]]
```
Comment (rekmarks) [inline run.sh:65]:
```suggestion
if [[ "$ADD" = "--add" ]];
```

File: run.sh [hunk 11]
```diff
@@ -4,33 +4,76 @@ set -x
...
+  then
```
Comment (rekmarks) [inline run.sh:66]:
```suggestion
if [[ "$ADD" = "--add" ]]; then
```

File: run.sh [hunk 12]
```diff
@@ -4,33 +4,76 @@ set -x
...
+
 git config user.name github-actions
 git config user.email github-actions@github.com
+git remote set-url origin "$REPO"
+
+branch_exists=$(git ls-remote --quiet --heads "$REPO" gh-pages)
 
-if git checkout --orphan gh-pages
+if [[ -n "$branch_exists" ]]
   then
```
Comment (rekmarks) [inline run.sh:53]:
```suggestion
if [[ -n "$branch_exists" ]]; then
```

File: run.sh [hunk 13]
```diff
@@ -4,33 +4,76 @@ set -x
...
+    echo "gh-pages branch already created"
+  else
```
Comment (rekmarks) [inline run.sh:55]:
```suggestion
else
```

File: action.yml [hunk 14]
```diff
@@ -1,19 +1,35 @@
...
   release-branch-prefix:
     description: 'The prefix used for branches of release PRs.'
     default: 'release/'
     required: true
+  destination-directory:
+    description: 'The directory within the GitHub Pages branch to place the source files.'
+    required: true
+    default: '.'
+  commit-message:
+    description: 'Commit Message'
```
Comment (rekmarks) [inline action.yml]:
```suggestion
    description: 'The messages of the commit to the gh-pages branch.'
```

**Review by rekmarks — APPROVED** [review]
LGTM!

**Pull request merged**